guidance/src/guidance/guidance_worker.cpp

```diff
@@ -0,0 +1,183 @@
+        for(int i = i; i < plugins.size(); ++i)
+        {
+            if(plugins[i].activated)
+            {
+                res.plugins.push_back(plugins[i]);
+            }
+        }
+    }
+
+    bool GuidanceWorker::is_required_plugin(std::string plugin_name, std::string version)
+    {
+        for(int i = 0; i < required_plugins.size(); i++)
+        {
+            if(plugin_name.append(" " + version).compare(required_plugins[i]) == 0)
+            {
+               return true;
+            }
+        }
+        return false;
+    }
+
+    bool GuidanceWorker::registered_plugin_cb(cav_srvs::PluginListRequest& req, cav_srvs::PluginListResponse& res)
+    {
+        populate_plugin_list_response(res);
+        return true;
+    }
+
+    bool GuidanceWorker::active_plugin_cb(cav_srvs::PluginListRequest& req, cav_srvs::PluginListResponse& res)
+    {
+        populate_active_plugin_list_response(res);
+        return true;
+    }
+
+    bool GuidanceWorker::activate_plugin_cb(cav_srvs::PluginActivationRequest& req, cav_srvs::PluginActivationResponse& res)
+    {
+        for(int i = 0; i < plugins.size(); ++i)
+        {
+            // Go through the plugin list to find the corresponding plugin that the user wants to activate
+            if(req.pluginName.compare(plugins[i].name) && req.pluginName.compare(plugins[i].versionId))
+            {
+                plugins[i].activated = req.activated;
+                res.newState = plugins[i].activated;
+                return true;
+            }
+        }
+        return false;
+    }
+
+    void GuidanceWorker::system_alert_cb(const cav_msgs::SystemAlertConstPtr& msg)
+    {
+        gsm.onSystemAlert(msg);
+    }
+
+    void GuidanceWorker::plugin_discovery_cb(cav_msgs::Plugin msg)
+    {
+        // if plugin list does not have anything, add it to the list and return
+        if(plugins.size() == 0)
+        {
+            // check if a plugin required when it is newly discovered
+            msg.required = is_required_plugin(msg.name, msg.versionId);
+            plugins.push_back(msg);
+            return;
+        }
+        // if we have some entries in the list, check if the plugin is already in the record
+        size_t index = 0;
+        while(index < plugins.size())
+        {
+            // check version id to be safe, but in the normal case we do not allow same plugin with different version id existing together
+            if(plugins[index].name.compare(msg.name) == 0 && plugins[index].versionId.compare(msg.versionId) == 0)
+            {
+                // only availability can be changed by plugin itself
+                plugins[index].available = msg.available;
+                // this break will make sure that "index" variable stays at the location of matching entry if it exists
+                break;
+            }
+            ++index;
+        }
+        // index points to the end of pllugin list without a matching entry
+        if(index == plugins.size())
```

Comment:
seems like you could just return rather break then this check wouldn't be nessasary